Fix combined DEPTH_STENCIL_ATTACHMENT binding in WebGL 2.0

lexaknyazev opened the pull request:
#### 70f7b05838c50bc578c505ddbab309df084939d3
<pre>
Fix combined DEPTH_STENCIL_ATTACHMENT binding in WebGL 2.0
<a href="https://bugs.webkit.org/show_bug.cgi?id=223337">https://bugs.webkit.org/show_bug.cgi?id=223337</a>

Reviewed by Darin Adler.

In WebGL 2.0 contexts, DEPTH_STENCIL_ATTACHMENT is considered
an alias for DEPTH_ATTACHMENT + STENCIL_ATTACHMENT.

Fixes forthcoming strengthened conformance test in
<a href="https://github.com/KhronosGroup/WebGL/pull/3455">https://github.com/KhronosGroup/WebGL/pull/3455</a>

* LayoutTests/TestExpectations:
* Source/WebCore/html/canvas/WebGLFramebuffer.cpp:
(WebCore::WebGLFramebuffer::removeAttachmentFromBoundFramebuffer):
(WebCore::WebGLFramebuffer::setAttachmentInternal):
(WebCore::WebGLFramebuffer::removeAttachmentInternal):

Canonical link: <a href="https://commits.webkit.org/252695@main">https://commits.webkit.org/252695@main</a>
</pre>

darinadler commented on Source/WebCore/html/canvas/WebGLFramebuffer.cpp:
```diff
@@ -865,7 +869,12 @@ void WebGLFramebuffer::setAttachmentInternal(GCGLenum attachment, GCGLenum texTa
     removeAttachmentInternal(locker, attachment);
     if (texture && texture->object()) {
-        m_attachments.set(attachment, WebGLTextureAttachment::create(texture, texTarget, level, layer));
+        RefPtr<WebGLAttachment> textureAttachment = WebGLTextureAttachment::create(texture, texTarget, level, layer);
```
This should use `auto`; the type should be `Ref` rather than `RefPtr`. Also no need to write out `RefPtr<WebGLAttachment>` in modern code.

lexaknyazev replied:
Done.

darinadler commented on Source/WebCore/html/canvas/WebGLFramebuffer.cpp:
```diff
@@ -865,7 +869,12 @@ void WebGLFramebuffer::setAttachmentInternal(GCGLenum attachment, GCGLenum texTa
+            m_attachments.set(GraphicsContextGL::DEPTH_ATTACHMENT, textureAttachment);
+            m_attachments.set(GraphicsContextGL::STENCIL_ATTACHMENT, textureAttachment);
+        }
+        m_attachments.set(attachment, textureAttachment);
```
This should use `WTFMove(textureAttachment)`.

lexaknyazev replied:
Semantically, the same `textureAttachment` reference needs to be stored under three different keys in the set. Using `WTFMove` moves it once and makes two others null.

darinadler replied:
That's incorrect and not how move semantics work. The first two times make additional references by copying the RefPtr. Using move the third time avoids reference count churn.

It's true that moving before reusing the value in the local variable would make the value null but that is not what I am suggesting.

lexaknyazev replied:
Thanks for the clarification wrt move. Using `auto` instead of `RefPtr` when declaring the variable causes a compile error:
```
~/WebKit/WebKitBuild/Release/usr/local/include/wtf/HashMap.h:373:32: error: no viable overloaded '='
        result.iterator->value = std::forward<V>(value);
        ~~~~~~~~~~~~~~~~~~~~~~ ^ ~~~~~~~~~~~~~~~~~~~~~~
~/WebKit/WebKitBuild/Release/usr/local/include/wtf/HashMap.h:396:12: note: in instantiation of function template specialization 'WTF::HashMap<unsigned int,
      WTF::RefPtr<WebCore::WebGLFramebuffer::WebGLAttachment>>::inlineSet<const unsigned int &, WTF::Ref<WebCore::WebGLFramebuffer::WebGLAttachment> &>' requested here
    return inlineSet(key, std::forward<T>(mapped));
           ^
In file included from ~/WebKit/WebKitBuild/Release/DerivedSources/WebCore/unified-sources/UnifiedSource193.cpp:1:
./html/canvas/WebGLFramebuffer.cpp:874:27: note: in instantiation of function template specialization 'WTF::HashMap<unsigned int,
      WTF::RefPtr<WebCore::WebGLFramebuffer::WebGLAttachment>>::set<WTF::Ref<WebCore::WebGLFramebuffer::WebGLAttachment> &>' requested here
            m_attachments.set(GraphicsContextGL::DEPTH_ATTACHMENT, textureAttachment);
                          ^
```

darinadler replied:
I believe the fix there is to write `textureAttachment.ptr()` instead of `textureAttachment` on those lines.

lexaknyazev replied:
Done.

darinadler commented on Source/WebCore/html/canvas/WebGLFramebuffer.cpp:
```diff
@@ -881,7 +890,12 @@ void WebGLFramebuffer::setAttachmentInternal(GCGLenum attachment, WebGLRenderbuf
     removeAttachmentInternal(locker, attachment);
     if (renderbuffer && renderbuffer->object()) {
-        m_attachments.set(attachment, WebGLRenderbufferAttachment::create(renderbuffer));
+        RefPtr<WebGLAttachment> renderbufferAttachment = WebGLRenderbufferAttachment::create(renderbuffer);
```
Same here.

darinadler commented on Source/WebCore/html/canvas/WebGLFramebuffer.cpp:
```diff
@@ -881,7 +890,12 @@ void WebGLFramebuffer::setAttachmentInternal(GCGLenum attachment, WebGLRenderbuf
+            m_attachments.set(GraphicsContextGL::DEPTH_ATTACHMENT, renderbufferAttachment);
+            m_attachments.set(GraphicsContextGL::STENCIL_ATTACHMENT, renderbufferAttachment);
+        }
+        m_attachments.set(attachment, renderbufferAttachment);
```
And here.

darinadler commented on LayoutTests/TestExpectations:
```diff
@@ -3795,6 +3795,7 @@ webgl/2.0.y [ Skip ]
 # Explicitly enable tests which we have fixed and do not have corresponding 2.0.0 test functionality.
 webgl/2.0.y/conformance/canvas/to-data-url-test.html [ Pass ]
 webgl/2.0.y/conformance/misc/invalid-passed-params.html [ Pass ]
+webgl/2.0.y/conformance/misc/expando-loss.html [ Pass ]
```
Does this test cover what is fixed here? Given the name of the test that seems surprising.

lexaknyazev replied:
This test iterates over different attachment points while re-binding renderbuffers and requesting their names (JS objects). Without this PR, calling `getFramebufferAttachmentParameter` with `DEPTH_STENCIL_ATTACHMENT` fails in WebGL 2.0 because the semantic of this attachment point is different than in WebGL 1.0.

lexaknyazev replied:
Filed https://github.com/KhronosGroup/WebGL/pull/3455.

darinadler reviewed (approved):
Still not satisfied with the test coverage here. Seems like I could come back here to change something, make a mistake and break this, and no one would notice.

lexaknyazev committed:
Fix combined DEPTH_STENCIL_ATTACHMENT binding in WebGL 2.0
https://bugs.webkit.org/show_bug.cgi?id=223337

Reviewed by Darin Adler.

In WebGL 2.0 contexts, DEPTH_STENCIL_ATTACHMENT is considered
an alias for DEPTH_ATTACHMENT + STENCIL_ATTACHMENT.

Fixes forthcoming strengthened conformance test in
https://github.com/KhronosGroup/WebGL/pull/3455

* LayoutTests/TestExpectations:
* Source/WebCore/html/canvas/WebGLFramebuffer.cpp:
(WebCore::WebGLFramebuffer::removeAttachmentFromBoundFramebuffer):
(WebCore::WebGLFramebuffer::setAttachmentInternal):
(WebCore::WebGLFramebuffer::removeAttachmentInternal):

Canonical link: https://commits.webkit.org/252695@main
